### PR TITLE
Transport system

### PR DESCRIPTION
**Pull request by reverendbedford**
Adds a simple 1D, finite volume transport solver to support a couple of projects.

### CODE REVIEW

File: tests/src/VTestVerification.cxx [hunk 1]
```diff
@@ -133,9 +142,16 @@ diff(const torch::jit::named_buffer_list & res,
...
+    Size j = 0;
     for (Size i = 0; i < nstep; i++)
     {
-      const auto refi = val.index({i}).squeeze();
+      if (!indices.empty() &&
+          !std::binary_search(indices.begin(), indices.end(), static_cast<size_t>(i)))
+        continue;
+
+      const auto refi = indices.size() == 1 ? at::Tensor(val) : val.index({j}).squeeze();
+      j++;
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply tests/src/VTestVerification.cxx:184]: It means the first option, but it absolutely doesn't skip verification entirely.  It just continues to i = 1, etc. until it finds the correct step.  I'll document better (and deal with the sort issue), but your logic is bad...

File: doc/ai-specifications/finite_volume.md
Comment (hugary1995) [inline doc/ai-specifications/finite_volume.md:1]: Is this the file you used to instruct copilot? Should we check it in?

File: doc/content/physics/finite_volume.md
Comment (hugary1995) [inline doc/content/physics/finite_volume.md:1]: I don't see this page in the documentation preview. I think you need to update the doxygen layouts to include this page.

File: doc/content/physics/finite_volume.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+\f[
+  u_t + J_x = R,
+\f]
+
+with
+
+\f[
+  J = J_{diffusion} + J_{advection}, \quad
+  J_{diffusion} = -D u_x, \quad
+  J_{advection} = v u.
```
Comment (hugary1995) [inline doc/content/physics/finite_volume.md]: I don't like italic subscripts :)

File: doc/content/physics/finite_volume.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+- **LinspaceScalar**: Uniform edge locations or time grids.
+- **CenterScalar**: Cell centers from edge locations.
+- **DifferenceScalar**: Cell sizes from edge locations.
+- **GaussianScalar**: Convenient Gaussian initial conditions.
+
+## Example: combined advection–diffusion–reaction
+
+Below is a compact example that assembles a full transport system, applies boundary conditions, and advances in time using backward Euler. The full regression test can be found in tests/regression/transport/combined/model.i.
+
+```
```
Comment (hugary1995) [inline doc/content/physics/finite_volume.md]: I added a markdown directive to list input file (without manual copying, which could be fragile). Maybe grep for `@list-hit` or something like that. I haven't got a chance to update the other physics documentation pages yet, but I will.

File: src/neml2/user_tensors/CenterTensor.cxx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+{
+  auto * f = this->factory();
+  neml_assert(f, "Failed assertion: factory != nullptr");
+
+  const auto points = _points.resolve(f);
+  const auto n = points.intmd_size(_dim);
+  const auto left = points.intmd_slice(_dim, indexing::Slice(0, n - 1));
+  const auto right = points.intmd_slice(_dim, indexing::Slice(1, n));
+  return 0.5 * (left + right);
+}
```
Comment (hugary1995) [inline src/neml2/user_tensors/CenterTensor.cxx:70]: This requires `n >= 2`, right?

File: include/neml2/user_tensors/CenterTensor.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// THE SOFTWARE.
+
+#pragma once
+
+#include "neml2/user_tensors/UserTensorBase.h"
+
+namespace neml2
+{
+/**
+ * @brief Compute interval centers along an intermediate dimension.
```
Comment (hugary1995) [inline include/neml2/user_tensors/CenterTensor.h:32]: We probably shouldn't hard-code this to always operate on intermediate dimensions. Take a look at `LinspaceTensor.cxx`, you can add a enum to control which dimension group to operate on.

Comment (hugary1995) [reply include/neml2/user_tensors/CenterTensor.h:32]: Same comment applies to `DifferenceTensor`.

File: src/neml2/user_tensors/GaussianTensor.cxx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace neml2
+{
+template <typename T>
+OptionSet
+GaussianTensorTmpl<T>::expected_options()
+{
+  OptionSet options = UserTensorBase<T>::expected_options();
+  options.doc() = "Construct a " + UserTensorBase<T>::tensor_type() +
+                  " with values sampled from a Gaussian profile at given points.";
```
Comment (hugary1995) [inline src/neml2/user_tensors/GaussianTensor.cxx]: It would help to include an equation in the description.

File: src/neml2/models/transport/CellRateOfChange.cxx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    _u_dot(declare_output_variable<Scalar>("rate"))
+{
+}
+
+void
+CellRateOfChange::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
+{
+  const auto dJ = intmd_diff(_flux(), 1, -1);
+  const auto N = _R.intmd_size(-1);
+  const auto dx_vec = (_dx.intmd_dim() == 0 ? _dx.intmd_expand(N) : _dx);
```
Comment (hugary1995) [inline src/neml2/models/transport/CellRateOfChange.cxx]: Why `intmd_expand` rather than relying on our broadcasting rule?

File: src/neml2/models/transport/CellRateOfChange.cxx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  const auto inv_dx = 1.0 / dx_vec;
+
+  if (out)
+    _u_dot = _R - dJ * inv_dx;
```
Comment (hugary1995) [inline src/neml2/models/transport/CellRateOfChange.cxx]: I wonder what we are saving by having such a specialized kernel. We have two things here, a reaction term, and a div(J). We could use a ScalarLinearInterpolation to add the two.

div(J) could be named much more general, something like FV1DDivergence -- as that's exactly what it does.

File: src/neml2/models/finite_volume/FiniteVolumeGradient.cxx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  const auto D_vec = (_D_edge.intmd_dim() == 0 ? _D_edge().intmd_expand(M) : _D_edge());
+  const auto x_vec =
+      (_cell_centers.intmd_dim() == 0 ? _cell_centers.intmd_expand(N) : _cell_centers);
+  const auto x_left = x_vec.intmd_slice(-1, indexing::Slice(0, N - 1));
+  const auto x_right = x_vec.intmd_slice(-1, indexing::Slice(1, N));
+
+  const auto dx = x_right - x_left;
+  const auto inv_dx = 1.0 / dx;
+
+  const auto du = u_right - u_left;
```
Comment (hugary1995) [inline src/neml2/models/finite_volume/FiniteVolumeGradient.cxx]: This is `intmd_diff` isn't it?

File: src/neml2/models/transport/DiffusiveFlux.cxx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  const auto M = N - 1;
+  const auto u_left = _u().intmd_slice(-1, indexing::Slice(0, N - 1));
+  const auto u_right = _u().intmd_slice(-1, indexing::Slice(1, N));
+  const auto D_vec = (_D_edge.intmd_dim() == 0 ? _D_edge().intmd_expand(M) : _D_edge());
+  const auto x_vec =
+      (_cell_centers.intmd_dim() == 0 ? _cell_centers.intmd_expand(N) : _cell_centers);
+  const auto x_left = x_vec.intmd_slice(-1, indexing::Slice(0, N - 1));
+  const auto x_right = x_vec.intmd_slice(-1, indexing::Slice(1, N));
+
+  const auto dx = x_right - x_left;
```
Comment (hugary1995) [inline src/neml2/models/transport/DiffusiveFlux.cxx]: This is `intmd_diff` isn't it?

File: src/neml2/models/transport/DiffusiveFlux.cxx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+void
+DiffusiveFlux::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
+{
+  const auto N = _u.intmd_size(-1);
+  const auto M = N - 1;
+  const auto u_left = _u().intmd_slice(-1, indexing::Slice(0, N - 1));
+  const auto u_right = _u().intmd_slice(-1, indexing::Slice(1, N));
+  const auto D_vec = (_D_edge.intmd_dim() == 0 ? _D_edge().intmd_expand(M) : _D_edge());
+  const auto x_vec =
+      (_cell_centers.intmd_dim() == 0 ? _cell_centers.intmd_expand(N) : _cell_centers);
```
Comment (hugary1995) [inline src/neml2/models/transport/DiffusiveFlux.cxx]: I'm not sure if the `intmd_expand`s are necessary.

File: src/neml2/models/transport/DiffusiveFlux.cxx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+  const auto x_left = x_vec.intmd_slice(-1, indexing::Slice(0, N - 1));
+  const auto x_right = x_vec.intmd_slice(-1, indexing::Slice(1, N));
+
+  const auto dx = x_right - x_left;
+  const auto inv_dx = 1.0 / dx;
+
+  const auto du = u_right - u_left;
+
+  if (out)
+    _J = -D_vec * du * inv_dx;
```
Comment (hugary1995) [inline src/neml2/models/transport/DiffusiveFlux.cxx]: After identifying the `intmd_diff`s, this kernel appears to have the exact same form as `FV1DDivergence` (or currently, `CellRateOfChange`). I guess in 1D, divergence and gradient do share the same form.

File: src/neml2/models/transport/LinearReaction.cxx [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    _u(declare_input_variable<Scalar>("u")),
+    _R(declare_output_variable<Scalar>("reaction"))
+{
+}
+
+void
+LinearReaction::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
+{
+  if (out)
+    _R = -_lambda * _u;
```
Comment (hugary1995) [inline src/neml2/models/transport/LinearReaction.cxx]: This is also just `ScalarLinearCombination`.

**Review by hugary1995 — CHANGES_REQUESTED** [review]
My main concern is code duplication. See my comments below.

A minor suggestion is to rename transport -> finite_volume. Most routines can be made general to any equation discretized using FV (not just transport equations). Also, it's unlikely that you'll include a different discretization method in the transport module, so perhaps just name it FV.

File: doc/ai-specifications/finite_volume.md
Comment (reverendbedford) [reply doc/ai-specifications/finite_volume.md:1]: It's now quite inaccurate so I removed it.

File: include/neml2/user_tensors/CenterTensor.h (hunk 5 above)
Comment (reverendbedford) [reply include/neml2/user_tensors/CenterTensor.h:32]: This one I'm leaving, as it drove me nuts when I had to specify each time like in Linspace...

**Review by hugary1995 — APPROVED** [review]
